debug.c

```diff
@@ -46,44 +51,170 @@
 
 static pid_t db_pid = -1;
 static pid_t shell_pid = -1;
+static struct timer shell_timer;
+static bool shell_session = false;
+static bool shell_timeout_active = false;
 
-void pv_debug_start_shell()
+static uint64_t pv_debug_timeout_elapsed_sec()
+{
+	struct timer_state state = timer_current_state(&shell_timer);
+
+	if (!state.fin)
+		return state.sec;
+
+	return 0;
+}
+
+bool pv_debug_start_shell()
 {
 	char c[64] = { 0 };
-	int t = 5;
 	int con_fd;
+	static bool notify_once = true;
 
-	if (shell_pid > -1)
-		return;
+	if (!pv_config_get_bool(PV_DEBUG_SHELL))
+		return false;
+
+	if (notify_once) {
+		pv_wall_welcome();
+		notify_once = false;
+	}
 
-	con_fd = open("/dev/console", O_RDWR);
+	if (shell_pid < 0) {
+		shell_session = false;
+	}
+
+	if (pv_config_get_bool(PV_DEBUG_SHELL_AUTOLOGIN)) {
+		shell_pid =
+			tsh_run("/sbin/getty -n -l /bin/sh 0 console", 0, NULL);
+		shell_session = true;
+		goto out;
+	}
+
+	con_fd = open("/dev/console", O_RDONLY | O_NONBLOCK | O_CLOEXEC );
 	if (con_fd < 0) {
 		pv_log(WARN, "Unable to open /dev/console");
-		return;
+		return false;
 	}
 
-	dprintf(con_fd, "Press [d] and [ENTER] for debug ash shell... ");
-	fcntl(con_fd, F_SETFL, fcntl(con_fd, F_GETFL) | O_NONBLOCK);
-	while (t && (read(con_fd, &c, sizeof(c)) < 0)) {
-		dprintf(con_fd, "%d ", t);
-		fflush(NULL);
-		sleep(1);
-		t--;
-	}
-	dprintf(con_fd, "\n");
+	read(con_fd, &c, sizeof(c));
+	close(con_fd);
 
-	if (c[0] == 'd' || pv_config_get_bool(PV_DEBUG_SHELL_AUTOLOGIN))
+	if (c[0] == '\n') {
 		shell_pid =
 			tsh_run("/sbin/getty -n -l /bin/sh 0 console", 0, NULL);
+
+		shell_session = true;
+		pv_wall("new shell session");
+		pv_log(INFO, "shell started with pid %d", shell_pid);
+		goto out;
+	}
+
+	return false;
+out:
+	return true;
 }
 
-void pv_debug_wait_shell()
+void pv_debug_stop_shell()
+{
+	if (shell_pid < 0)
+		return;
+
+	sigset_t old_sigset;
+	int status = 0;
+	if (pvsignals_block_chld(&old_sigset) != 0) {
+		pv_log(WARN, "Failed to block SIGCHLD");
+		return;
+	}
+
+	if (!kill(shell_pid, SIGKILL)) {
+		waitpid(shell_pid, &status, 0);
+	}
+
+	if (pvsignals_setmask(&old_sigset) != 0) {
+		pv_log(WARN, "Failed to restore signal mask");
+	}
+
+	shell_pid = -1;
+
+	shell_session = false;
+	pv_wall("ended shell session");
+}
+
+bool pv_debug_start_timeout_shell()
 {
-	if (shell_pid > -1) {
-		pv_log(WARN, "waiting for debug shell with pid %d to exit",
-		       shell_pid);
-		waitpid(shell_pid, NULL, 0);
+	int debug_timeout = 0;
+
+	if(shell_timeout_active) {
+		return false;
+	}
+
+	if (pv_config_get_bool(PV_DEBUG_SHELL)) {
```

Comment:
I guess you have to guard this with shell_session too? Otherwise, we will enter here as long as PV_DEBUG_SHELL is 1 and even if no shell is opened. Instead of this big indented thingy, I'd do something like:

```
if (!pv_config_get_bool(PV_DEBUG_SHELL) || !shell_session)
    return false;
    
// rest of the stuff
```